falcon/cyutil/uri.pyx

```diff
@@ -0,0 +1,299 @@
+            else:
+                result[key] = [decode(element) for element in value.split(',')]
+        else:
+            result[key] = decode(value)
+
+    elif isinstance(old_value, list):
+        old_value.append(decode(value))
+    else:
+        result[key] = [old_value, decode(value)]
+
+
+cdef cy_parse_query_string(unsigned char* data, Py_ssize_t length,
+                           bint keep_blank, bint csv):
+    cdef Py_ssize_t pos
+    cdef unsigned char current
+
+    cdef Py_ssize_t start = 0
+    cdef Py_ssize_t encoded_start_key = -1
+    cdef Py_ssize_t encoded_start_val = -1
+    cdef Py_ssize_t partition = -1
+
+    cdef unicode key
+    cdef unicode value
+    cdef old_value
+    cdef dict result = {}
+
+    for pos in range(length):
+        # PERF(vytas): Quick check if we need to do anything special with the
+        #   current character.
+        #   Cython should translate this check into a switch statement.
+        if data[pos] not in b'%&+,=':
+            continue
+
+        current = data[pos]
+
+        if current == b'&':
+            if pos > start:
+                if partition >= 0:
+                    key = cy_decode(data, start, partition, encoded_start_key)
+                    if csv and encoded_start_val >= 0:
+                        cy_handle_csv(result, keep_blank, key,
+                                      data[partition+1:pos].decode())
+                        start = pos + 1
+                        encoded_start_key = -1
+                        encoded_start_val = -1
+                        partition = -1
+                        continue
+
+                    value = cy_decode(data, partition+1, pos, encoded_start_val)
+                else:
+                    key = cy_decode(data, start, pos, encoded_start_key)
+                    value = EMPTY_STRING
+
+                if value is not EMPTY_STRING or keep_blank:
+                    old_value = result.get(key)
+
+                    if old_value is None:
+                        result[key] = value
+                    elif isinstance(old_value, list):
+                        old_value.append(value)
+                    else:
+                        result[key] = [old_value, value]
+
+            start = pos + 1
+            encoded_start_key = -1
+            encoded_start_val = -1
+            partition = -1
+            continue
+
+        if current == b'=':
+            if partition < 0:
+                partition = pos
+            continue
+
+        # else: current in b'%+,'
+
+        # PERF(vytas): Record positions of the first encoded character, if any.
+        #  This will be used to determine where to start decoding, if at all.
+        if partition < 0:
+            if encoded_start_key < 0:
+                encoded_start_key = pos
+        else:
+            if encoded_start_val < 0:
+                encoded_start_val = pos
+
+    if length > start:
+        if partition >= 0:
```

Comment:
I know that it would probably require passing a long list of values to factor this out to a function, or passing in a struct-like state object, so it may not be worth DRYing this up. However, it may be worth prefixing this section and the section above with a comment reminding the reader to keep the code in the two locations in sync.

Reply:
Agreed, this block grew a bit copypaste-ish. I'll see what I can do about it. In the worst case I'll add some comments indeed.